### PR TITLE
lapack/gonum: add Dgesc2

### PR DESCRIPTION
**Pull request by soypat**
So the checklist is NOT complete. I intend for this to be a first review of how LAPACK routines are implemented in Go and get some feedback. I'm sure my indexing is messed up, was quite confused seeeing how fortran alternated betweeen double and single indexing.

No tests implemented.

### CODE REVIEW

File: lapack/gonum/idamax.go
Comment (kortschak) [inline lapack/gonum/idamax.go:16]: `Idamax` is [implemented in BLAS level 1](https://pkg.go.dev/gonum.org/v1/gonum/blas#Float64Level1), so this file is not needed.

File: lapack/gonum/dlabad.go
Comment (kortschak) [inline lapack/gonum/dlabad.go:17]: As mentioned by @vladimir-ch we don't use `Dlabad` since we use IEEE 754, so this file can be deleted.

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:13]: `Dgesc2`? and rename file?

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:19]: `const smlnum = dlamchS/dlamchP`

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:23]: `// Apply permutations ipiv to RHS.` (final dot and space between comment marks to distinguish from directives). Though I'm not sure if this comment is needed (@vladimir-ch?).

Also separate this comment from the conditional panic by a blank line.

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:26]: `// Solve for L part.`

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:27]: `i < n-1`?

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:31]: `rhs[j] -= float64(a[j*lda + i]*rhs[i])`

Yes, this is currently optimised for column-major storage.

Apparent unnecessary conversion is there to prevent FMS on arm64.

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:35]:
```
<blank line>
// Solve for U part.

scale = 1.0

// Check for scaling.
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:36]:
```
bi := blas64.Implementation()
i := bi.Iamax(blas64.Vector{N: n, Data: rhs, Inc: 1})
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:37]: `math.Abs(a[n*lda+n])`

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:39]:
```
bi.Scal(rhs[0], blas64.Vector{N: n, Data: temp, Inc: 1})
```

This is scaling `temp` by the first element of `rhs`. The blas64 API takes a vector type here.

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:40]:
```
scale *= temp
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:42]:
```
for i := n-1; i >= 0; i-- {
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:43]:
```
temp := 1 / a[i*lda + i]
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:44]:
```
rhs[i] *= temp
```

File: lapack/gonum/dgesdc2.go
Comment (kortschak) [inline lapack/gonum/dgesdc2.go:46]:
```
rhs[i] -= float64(rhs[j]*a[i*lda + j]*temp)
```

File: lapack/gonum/dgesdc2.go
Comment (soypat) [reply lapack/gonum/dgesdc2.go:27]: Right, I see it now. Thanks!

File: lapack/gonum/dgesdc2.go
Comment (soypat) [reply lapack/gonum/dgesdc2.go:36]: Is this possible? I've found `Iamax` in namespace which calls Idamax and Idamax in Implementation:
```go
// one way.
bi := blas64.Implementation()
i := bi.Idamax(n,rhs,1)
// second way i managed.
i := blas64.Iamax(blas64.Vector{N: n, Data: rhs, Inc: 1})
```
Which one?

Comment (kortschak) [reply lapack/gonum/dgesdc2.go:36]: The first one since we already know `Inc` >= 0.

File: lapack/gonum/dgesdc2.go
Comment (soypat) [reply lapack/gonum/dgesdc2.go:39]: Again I'm getting the same problem where `blas64.Scal` has the signature you describe while `bi` has `Dscal` available.

Second observation: `temp` is not an array (or is it??). I'm seeing it the other way around, a part of `rhs` (a single element?) is being scaled by `temp`.  Now I'm looking at it more closely, I think it's scaling `rhs[0:]`. Please confirm.

Comment (kortschak) [reply lapack/gonum/dgesdc2.go]: > Again I'm getting the same problem where `blas64.Scal` has the signature you describe while `bi` has `Dscal` available.

Sorry, `bi.Dscal`.

> `temp` is not an array (or is it??). I'm seeing it the other way around, a part of `rhs` (a single element?) is being scaled by temp. Now I'm looking at it more closely, I think it's scaling `rhs[0:]`. Please confirm.

Again, sorry. You are right. `temp` is a scalar and `rhs(1)` is being passed to be scaled.

File: lapack/gonum/dgesc2.go
Comment (kortschak) [inline lapack/gonum/dgesc2.go:17]: Please add that the result is placed in `rhs` on exit.

Also, since this is an auxiliary routine, it's worth adding
```
//
// Dgesc2 is an internal routine. It is exported for testing purposes.
```

File: lapack/gonum/dgesc2.go
Comment (kortschak) [inline lapack/gonum/dgesc2.go:26]: Uppercase 'S' in solve.

File: lapack/gonum/dgesc2.go
Comment (kortschak) [inline lapack/gonum/dgesc2.go:14]: Nit: clean up the equation, `//  A * X = scale * RHS` (two spaces between comment marks and text and single spaces around the operators).

**Review by kortschak — COMMENTED** [review]
This looks good now. Thanks. Some minor nits and tests (take a look at how @vladimir-ch implements tests in the testlapack package) — here you would take random inputs with a variety of dimensions and strides, do the LU with `Dgetc2` (needs to be implemented - looks very straightforward) and pass in the LU part, checking on return that the equation in the docs is satisfied.

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgesc2.go:50]: No need for an explicit conversion and please keep the parenthesis as in the reference.

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgesc2.go:55]: `lda` must be `1`, 4. and 5. parameter `0` and `n-2`, and `jpiv` must be sliced to `n-1`.

**Review by vladimir-ch — CHANGES_REQUESTED** [review]
This needs a test before merging for which `Dgetc2` must be implemented (with a test).

File: lapack/gonum/dgesc2.go
Comment (kortschak) [reply lapack/gonum/dgesc2.go:29]: This apparently unnecessary conversion was added at my request to prevent FMS generation on arm64, same below.

Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:29]: Ah, ok, I saw your comment in email but didn't relate it to this code when reviewing.

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:24]: I thought it's strange that `k2` is `n-1` and not `n` in the Fortran reference. In fact it's because they rely on an undocumented implementation detail of DGETC2:
```
*
*     Set last pivots to N
*
      IPIV( N ) = N
      JPIV( N ) = N
```
I think it would be cleaner to pass `0, n-1, ipiv[:n], 1` here. What do you think, @kortschak ?

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:55]: The same here: `0, n-1, jpiv[:n], -1`.

File: lapack/gonum/dgesc2.go
Comment (kortschak) [reply lapack/gonum/dgesc2.go:24]: Wouldn't that mean were doing an additional row of work that is in effect a no-op?

Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:24]: Yes but only because our `Dlaswp` implementation doesn't check whether it's doing a no-op swap (unlike the reference which does the check). I think it would be better to pass `n` pivots here and below and change `Dlaswp` to check for a no-op.

Comment (kortschak) [reply lapack/gonum/dgesc2.go:24]: Yeah, OK.

**Comment by soypat** [comment]
Some lessons learned:
* Parenthesis matter. They can prevent precision loss and overflow if done correctly.
* `lda` is the "leading dimension" and used to index matrices, not `n`/`m` (actual dimensions)
* Documentation must be taken care of properly
* Add bounds/length checks not present in reference

As for the test, I need to generate not only the general matrix but also `ipiv` and `jpiv`. I'll take a look at other tests which do this i.e. `dlaqp2.go` and `dlaqps.go`

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgesc2.go:21]: This part should look like this:
```
	switch {
	case n < 0:
		panic(nLT0)
	case lda < max(1, n):
		panic(badLdA)
    }

    // Quick return if possible.
    if n == 0 {
        return 0
    }

    switch {
	case len(a) < (n-1)*lda+n:
		panic(shortA)
	case len(rhs) < n:
		panic(shortRHS)
	case len(ipiv) != n:
	    panic(badLenIpiv)
	case len(jpiv) != n:
	    panic(badLenJpiv)
	}

	const smlnum = dlamchS / dlamchP
```

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:38]: #1654 submitted

File: lapack/gonum/dgesc2.go
Comment (soypat) [reply lapack/gonum/dgesc2.go:21]: I've applied your suggestions. I'm still confused as to how much we have to test. Is it enough that it did not panic with a bunch of random matrices? Should the result be compared with another linear solver?

**Comment by vladimir-ch** [comment]
> As for the test, I need to generate not only the general matrix but also ipiv and jpiv.

It will be much easier (and more correct) if you first implement Dgetc2 and use it in the test to generate input for Dgesc2.

**Comment by soypat** [comment]
> It will be much easier (and more correct) if you first implement Dgetc2 and use it in the test to generate input for Dgesc2.

should I do it in this PR? should I start another?

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:21]: We must test that, given the output from Dgetc2, Dgesc2 computes what it promises, that is, that:
```
A * X = scale * RHS
```
holds and that `scale` is in `[0,1]`.

I would start with:
1. Generate a random A, compute the norm |A|
2. Generate a random X, compute the norm |X|
3. Compute RHS from A*X
4. Call Dgetc2
5. Call Dgesc2
6. Check that scale is in [0,1]
7. Check that |A*X - scale*RHS| / |A| / |X| is small

If nothing special is done, the above steps will always return `scale = 1`. Then the next step would be to modify 1. and/or 2. to trigger a non-unit scaling.

**Comment by vladimir-ch** [comment]
`Dgetc2` should be a separate PR.

**Comment by soypat** [comment]
Dgesc2 is passing albeit without the test you mentioned. I tried programming it and ended up leaving the `t.Error` declaration commented because it was failing although `X` was being calculated correctly.

**Comment by soypat** [comment]
*pinging*. Needed for https://github.com/gonum/gonum/pull/1658 and #1669.

worth mentioning https://github.com/gonum/gonum/pull/1665 is ready for merge too.

File: lapack/testlapack/dgesc2.go
Comment (vladimir-ch) [inline lapack/testlapack/dgesc2.go:1]: 2021

File: lapack/gonum/dgesc2.go
Comment (vladimir-ch) [inline lapack/gonum/dgesc2.go:15]: `n×n`

**Review by vladimir-ch — COMMENTED** [review]
LGTM after the two comments. I'll fix the test later.

**Comment by soypat** [comment]
@vladimir-ch Thanks for taking a look!

**Review by vladimir-ch — APPROVED** [review]

File: lapack/gonum/dgesc2.go
Comment (soypat) [inline lapack/gonum/dgesc2.go:50]: @kortschak or @vladimir-ch , I went back to `Dgesc2` to see how I used `Dlaswp` and this is confusing me some. The reference material does `CALL dlaswp( 1, rhs, lda, 1, n-1, ipiv, 1 )`, which according to its documentations, does row swap for `k2-k1+1` rows, that is **n-1** rows. Are we not swapping **n** rows here? Or am I misunderstanding?

Comment (vladimir-ch) [reply lapack/gonum/dgesc2.go:50]: This was discussed in https://github.com/gonum/gonum/pull/1652#discussion_r646106943. The reference documents IPIV and JPIV to have length n, DGETC2 sets n elements but the only users of the PIV arrays, DGESC2 and DLATDF, refer only to the first n-1 elements. That's because n-1 swaps are enough to represent the permutation matrix from DGETC2 and from the way PIV arrays are constructed the last element must be n-1. DGESC2 uses this fact and does not call DLASWP with the last no-op swap.

I see three options:
1. Keep what we have: ipiv and jpiv have length n, Dgetc2 sets all elements, Dgesc2 and Dlatdf pass all n elements to Dlaswp, Dlaswp will not do anything for the last swap.
2. Do what the reference does: ipiv and jpiv have length n, Dgetc2 sets all elements but Dgesc2 and Dlatdf pass only n-1 elements to Dlaswp
3. ipiv and jpiv have length n-1, Dgetc2 sets all elements, Dgesc2 and Dlatdf pass all n-1 elements to Dlaswp.

I think what the reference does is the least clear of these options (set n elements but then use only n-1 elements for no documented, albeit logical, reason). Option 1 or option 3 are both fine, option 1 diverges less from the reference.

Comment (soypat) [reply lapack/gonum/dgesc2.go:50]: Thank you for clearing it up. I remember you and Dan talked about it but I didn't really understand what the fuss was all about. I do now, crystal clear explanation, btw!

**Pull request merged**
